xl-v1-metadata.go

```diff
@@ -352,38 +375,23 @@ func (xl xlObjects) writeUniqueXLMetadata(bucket, prefix string, xlMetas []xlMet
 	// Wait for all the routines.
 	wg.Wait()
 
-	var errCount = 0
-	// Return the first error.
-	for _, err := range mErrs {
-		if err == nil {
-			continue
-		}
-		errCount++
-	}
-	// Count all the errors and validate if we have write quorum.
-	if errCount > len(xl.storageDisks)-xl.writeQuorum {
+	if !isQuorum(mErrs, xl.writeQuorum) {
 		// Validate if we have read quorum, then return success.
-		if errCount > len(xl.storageDisks)-xl.readQuorum {
+		if isQuorum(mErrs, xl.readQuorum) {
 			return nil
 		}
-		// Delete all the `xl.json` left over.
-		for index, disk := range xl.storageDisks {
-			if disk == nil {
-				continue
-			}
-			// Undo rename object in parallel.
-			wg.Add(1)
-			go func(index int, disk StorageAPI) {
-				defer wg.Done()
-				if mErrs[index] != nil {
-					return
-				}
-				_ = deleteXLMetdata(disk, bucket, prefix)
-			}(index, disk)
-		}
-		wg.Wait()
+		xl.deleteAllXLMetadata(bucket, prefix, mErrs)
 		return errXLWriteQuorum
 	}
+
+	for _, err := range mErrs {
+		if err != nil && err != errDiskNotFound {
+			xl.deleteAllXLMetadata(bucket, prefix, mErrs)
```

Comment:
> since writeUniqueXLMetadata writes to the tmp location doing xl.deleteAllXLMetadata is not needed

The function itself has no such awareness of tmp location so that is why it is done this way.